src/mappings/handlers/multisigEventHandler.ts

```diff
@@ -0,0 +1,157 @@
+
+  const operationId = generateOperationId(callHashString, multisigAccountId, blockCreated, indexCreated);
+
+  const newOperation = await MultisigOperation.create({
+    ...existingOperation,
+    id: operationId,
+    callHash: callHashString,
+    status: existingOperation?.status || OperationStatus.pending,
+    accountId: multisigAccountId,
+    depositor: u8aToHex(decodeAddress(event.extrinsic.extrinsic.signer.toString())),
+    blockCreated: blockCreated,
+    indexCreated: indexCreated,
+    timestamp: timestamp(event.extrinsic.block),
+  });
+
+  await newOperation.save();
+
+  const handleCall = async (visitedCall: VisitedCall) => {
+    const call = getDataFromCall<CallBase<AnyTuple>>(visitedCall.call, "call");
+
+    const existingOperation = await findExistingOperation(callHashString, blockCreated, indexCreated, multisigAccountId);
+
+    if (!existingOperation) throw new Error("Operation not found and that is fucking bad");
+
+    let callData = call?.toHex() || existingOperation?.callData;
+    let method = call?.method || existingOperation?.method;
+    let section = call?.section || existingOperation?.section;
+
+    const updatedOperation = await MultisigOperation.create({
+      ...existingOperation,
+      callData: callData,
+      method: method,
+      section: section,
+    });
+    await updatedOperation.save();
+
+    const signer = u8aToHex(decodeAddress(visitedCall.origin));
+
+    await MultisigEvent.create({
+      id: generateEventId(updatedOperation.id, signer, EventStatus.approve),
+      accountId: signer,
+      status: EventStatus.approve,
+      blockCreated: getBlockCreated(visitedCall.extrinsic),
+      indexCreated: getIndexCreated(visitedCall.extrinsic),
+      multisigId: updatedOperation.id,
+      timestamp: timestamp(visitedCall.extrinsic.block),
+    }).save();
+  };
+
+  const multisigVisitorHuy = CreateCallVisitorBuilder()
+    .on("utility", ["batch", "batchAll", "forceBatch"], (extrinsic, context) => {
+      const calls = extrinsic.call.args.at(0);
+      if (Array.isArray(calls) && calls.length > 100) {
+        // we're skipping large batches, something terrible happens inside anyway
+        context.stop();
+      }
+    })
+    .on("multisig", "asMulti", handleCall)
+    .on("multisig", "asMultiThreshold1", handleCall)
+    .on("multisig", "approveAsMulti", handleCall)
+    .on("multisig", "cancelAsMulti", handleCall)
+    .ignoreFailedCalls(true)
+    .build();
+
+  await callWalk.walk(event.extrinsic, multisigVisitorHuy);
+
+  logger.info(`Multisig event method: ${event.event.method}`);
+
+  const finalEvent = event.event.method === "MultisigExecuted" || event.event.method === "MultisigCancelled";
+  if (!finalEvent) return;
+
+  const result = getDataFromEvent<DispatchResult>(event.event, "result", 4);
+
+  let status = result?.isOk ? OperationStatus.executed : OperationStatus.error;
+  if (event.event.method === "MultisigCancelled") {
+    status = OperationStatus.cancelled;
+  }
```

Comment:
To add up to the comment above:
Setting status to approved to later replace it to cancelled in case we are processing cancelled even is extremely confusing - I think separate handlers will greatly improve the readability here